lib/rouge/lexers/ecl.rb

```diff
@@ -0,0 +1,152 @@
+          fileposition forward grouped inner internal linkcounted literal lzw 
+          mofn multiple namespace wnotrim noxpath onfail prefetch retry 
+          rowset scope smart soapaction stable timelimit timeout unordered 
+          unstable update use width
+        )
+      end
+
+      def self.functions
+        @functions ||= Set.new %w(
+          abs acos aggregate allnodes apply ascii asin asstring 
+          atan _token ave case catch choose choosen choosesets 
+          clustersize combine correlation cos cosh count covariance cron 
+          dataset dedup define denormalize dictionary distribute distributed distribution 
+          ebcdic enth error evaluate event eventextra eventname exists 
+          exp failcode failmessage fetch fromunicode fromxml getenv getisvalid 
+          global graph group hash hashcrc having httpcall httpheader 
+          if iff index intformat isvalid iterate join keyunicode 
+          length library limit ln local log loop map 
+          matched matchlength matchposition matchtext matchunicode max merge mergejoin 
+          min nofold nolocal nonempty normalize parse pipe power 
+          preload process project pull random range rank ranked 
+          realformat recordof regexfind regexreplace regroup rejected rollup round 
+          roundup row rowdiff sample set sin sinh sizeof 
+          soapcall sort sorted sqrt stepped stored sum table 
+          tan tanh thisnode topn tounicode toxml transfer transform 
+          trim truncate typeof ungroup unicodeorder variance which workunit 
+          xmldecode xmlencode xmltext xmlunicode apply assert build buildindex 
+          evaluate fail keydiff keypatch loadxml nothor notify output 
+          parallel sequential soapcall wait
+        )
+      end
+
+      class1 = %w(
+        file date str math metaphone metaphone3 uni audit blas system
+      )
+
+      class2 = %w(
+        debug email job log thorlib util workunit
+      )
+
+      state :single_quote do
+        rule %r([xDQUV]?'([^'\\]*(?:\\.[^'\\]*)*)'), Str::Single
+        rule %r/\\(x\\h{2}|[0-2][0-7]{,2}|3[0-6][0-7]?|37[0-7]?|[4-7][0-7]?|.|$)/, Text
+      end
+
+      state :inline_whitespace do
+        rule %r/[ \t\r]+/, Text
+        rule %r/\\\n/, Text # line continuation
+        rule %r(/[*].*?[*]/)m, Comment::Multiline
+      end
+
+      state :whitespace do
+        rule %r/\n+/m, Text
+        rule %r(//.*), Comment::Single
+        mixin :inline_whitespace
+      end
+
+      typenum = /${id}\d+/
+
+      state :root do
+        mixin :whitespace
+        mixin :single_quote
+
+        rule %r(\b(?i:(and|not|or|in))\b), Operator::Word
+        rule %r([:=|>|<|<>|/|\\|+|-|=]), Operator 
+        rule %r([\[\]{}();,\&,\.,\%]), Punctuation
+
+        rule %r(\b(?i:(begin|end)c\+\+)), Keyword
+        rule %r(\b(?i:(std)\.(#{class1.join('|')})\.(#{class2.join('|')}))\b), Name::Class
```

Comment:
Just out of curiosity, it seems based on the lexer that you can't define your own classes; there are simply the ones defined in the language. Does that mean `A.B.C` is invalid syntax if it doesn't match the `std` classes?

Reply:
The language doesn't support classes as such, but it does support nested constructs.  So A.B.C could be valid, but should really only be treated as a variable name.
`Std` is our builtin standard library so gets some special treatment.